src/marshmallow/utils.py

```diff
@@ -242,7 +242,9 @@ def _get_value_for_key(obj, key, default):
 
     try:
         return obj[key]
-    except (KeyError, IndexError, TypeError, AttributeError):
+    except (KeyError, IndexError, TypeError, AttributeError) as e:
+        if isinstance(e, KeyError) and type(obj) is dict:
```

Comment:
I'm not a fan of carving out an exception for `dict` from this dict-like class handling logic. [This method is a known hot path for serialization](https://github.com/marshmallow-code/marshmallow/issues/805#issuecomment-531019570). Would you consider creating an issue for this? I think there are solutions to this problem that would give users more direct control over this behavior and provide fast paths for serialization at the same time. Maybe something like `schema.dump(obj, from=dict)`.

Reply:
Opened an issue #1559